amaranth/lib/data.py

```diff
@@ -166,6 +169,9 @@ def as_shape(self):
         """
         return unsigned(self.size)
 
+    def __hash__(self):
+        return hash(self.size)
```

Comment:
This doesn't feel like a reasonable implementation.

Reply:
I guess what is reasonable depends on usage, one could add in a hash of `frozenset(iter(self))`, but that is quite slow. I assumed that most uses of `__hash__` would be having a `dict` with `Layout`s of different sizes, but maybe that is a bad assumption?

Reply:
Yes, precomputing the hash on creation.

Reply:
What about just putting `@functools.cache` on it?

Reply:
Also works, yeah!